src/ucp/core/ucp_context.c

```diff
@@ -693,32 +741,52 @@ ucs_status_t ucp_init_version(unsigned api_major_version, unsigned api_minor_ver
               context->num_mds, context->num_tls, context->config.features);
 
     *context_p = context;
-    return UCS_OK;
+    status     = UCS_OK;
+    goto out;
 
 err_free_config:
     ucp_free_config(context);
 err_free_ctx:
     ucs_free(context);
+out:
+    UCP_CONTEXT_THREAD_CS_EXIT(context);
 err:
     return status;
 }
 
 void ucp_cleanup(ucp_context_h context)
 {
+    UCP_CONTEXT_THREAD_CS_ENTER(context);
```

Comment:
no point to take mutex here, context is destroyed anyway, and there is no protection against that